### ISSUE TITLE
Moco_lisht

### ISSUE THREAD

**Issue by Nutty233** [Nutty233/Moco_learn#1]
import copy
import time

import torch
import torchvision
from torch import nn
from torch.utils.data import DataLoader
import os
import matplotlib.pyplot as plt
import numpy as np

from lightly.loss import NTXentLoss
from lightly.data import LightlyDataset
from lightly.models.modules import MoCoProjectionHead
from lightly.models.utils import deactivate_requires_grad, update_momentum
from lightly.transforms.moco_transform import MoCoV2Transform
from lightly.utils.scheduler import cosine_schedule

class MoCo(nn.Module):
    def __init__(self, backbone):
        super().__init__()

        self.backbone = backbone
        # self.projection_head = MoCoProjectionHead(512, 512, 128)
        self.projection_head = MoCoProjectionHead(25088, 512, 128)

        self.backbone_momentum = copy.deepcopy(self.backbone)
        self.projection_head_momentum = copy.deepcopy(self.projection_head)

        deactivate_requires_grad(self.backbone_momentum)
        deactivate_requires_grad(self.projection_head_momentum)

    def forward(self, x):
        query = self.backbone(x).flatten(start_dim=1)
        query = self.projection_head(query)
        return query

    def forward_momentum(self, x):
        key = self.backbone_momentum(x).flatten(start_dim=1)
        key = self.projection_head_momentum(key).detach()
        return key

def adjust_learning_rate(optimizer, epoch, start_lr):
    """Sets the learning rate to the initial LR decayed by 10 every 30 epochs"""
    lr = start_lr * (0.5 ** (epoch // 30))
    for param_group in optimizer.param_groups:
        param_group['lr'] = lr

def accuracy(output, target, topk=(1,)):
    """Computes the accuracy over the k top predictions for the specified values of k"""
    with torch.no_grad():
        maxk = max(topk)
        batch_size = target.size(0)

        _, pred = output.topk(maxk, 1, True, True)
        pred = pred.t()
        correct = pred.eq(target.view(1, -1).expand_as(pred))

        res = []
        for k in topk:
            correct_k = correct[:k].view(-1).float().sum(0, keepdim=True)
            res.append(correct_k.mul_(100.0 / batch_size))
        return res

if __name__ == '__main__':
    epochs = 600
    min_loss = 0
    ori_lr = 0.01
    lr_factor = 0.9
    train_losses = []
    pre_num = 0

    # ResNet
    # resnet = torchvision.models.resnet18()
    # backbone = nn.Sequential(*list(resnet.children())[:-1])

    # VGG16
    vgg16 = torchvision.models.vgg16()
    backbone = nn.Sequential(*list(vgg16.children())[:-1])
    model = MoCo(backbone)

    weight_dir = r'D:\...'
    # load weight if exist
    weight_best = os.path.join(weight_dir, f'moco_v2_loss_best.pt')
    if os.path.exists(weight_best):
        model.load_state_dict(torch.load(weight_best))
        train_losses = np.load(os.path.join(weight_dir, "train_losses.npy"))
        pre_num = len(train_losses)
        train_losses = list(train_losses)
        print(f'successful load weight!')
        print(f'already trained {pre_num:>03} epochs')

    device = "cuda" if torch.cuda.is_available() else "cpu"
    model.to(device)

    transform = MoCoV2Transform(input_size=224,
                                normalize={"mean": [0.1073, 0.1204, 0.1182, 0.2055], "std": [0.1051, 0.1109, 0.1153, 0.1599]})
    # dataset = torchvision.datasets.CIFAR10("data/test", download=True, transform=transform)
    # or create a dataset from a folder containing images or videos:
    # dataset = LightlyDataset("path/to/folder", transform=transform)
    dataset = LightlyDataset(r"D:\...",
                             transform=transform)

    dataloader = DataLoader(
        dataset,
        # batch_size=256,
        batch_size=128,
        shuffle=True,
        drop_last=True,
        # drop_last=False,
        num_workers=8,
    )

    criterion = NTXentLoss(memory_bank_size=4096)
    optimizer = torch.optim.SGD(model.parameters(), lr=ori_lr)
    scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer, 'min', factor=lr_factor, patience=20)

    print("Starting Training")
    for epoch in range(pre_num, epochs):
        adjust_learning_rate(optimizer, epoch, ori_lr)
        start_time = time.time()
        total_loss = 0
        momentum_val = cosine_schedule(epoch, epochs, 0.996, 1)
        for batch in dataloader:
            x_query, x_key = batch[0]
            update_momentum(model.backbone, model.backbone_momentum, m=momentum_val)
            update_momentum(
                model.projection_head, model.projection_head_momentum, m=momentum_val
            )
            x_query = x_query.to(device)
            x_key = x_key.to(device)
            query = model(x_query)
            key = model.forward_momentum(x_key)
            loss = criterion(query, key)
            total_loss += loss.detach()
            loss.backward()
            # lr = ori_lr * (0.5 ** (epoch // 40))
            # optimizer = torch.optim.SGD(model.parameters(), lr=lr)
            optimizer.step()
            optimizer.zero_grad()

        avg_loss = total_loss / len(dataloader)
        if epoch == 0:
            min_loss = avg_loss
            torch.save(model.state_dict(), os.path.join(weight_dir, f'moco_v2_loss_best.pt'))
        spend_time = time.time() - start_time
        print(f"epoch: {epoch:>03}, loss: {avg_loss:.5f}, spend: {spend_time:.2f}s, Lr:{optimizer.state_dict()['param_groups'][0]['lr']:.5f}")

        scheduler.step(avg_loss)

        train_loss = avg_loss.cpu()
        train_losses.append(train_loss)

        # save the best model
        if train_loss < min_loss:
            torch.save(model.state_dict(), os.path.join(weight_dir, f'moco_v2_loss_best.pt'))
            min_loss = train_loss
        # save the last model
        torch.save(model.state_dict(), os.path.join(weight_dir, f'moco_v2_last.pt')) 
        
    os.system("shutdown -s -t 60 ")
